src/pluto/mask.cr

```diff
@@ -0,0 +1,177 @@
+  end
+
+  private def resolve_to_start_and_count(range, size) : Tuple(Int32, Int32)
+    start, count = Indexable.range_to_index_and_count(range, size) || raise IndexError.new("Unable to resolve range #{range} for mask dimension of #{size}")
+    raise IndexError.new("Range #{range} exceeds bounds of #{size}") if (start + count) > size
+    {start, count}
+  end
+
+  def [](x : Int32, y : Int32) : Bool
+    raise IndexError.new("Out of bounds: this mask is #{width}x#{height}, and (#{x},#{y}) is outside of that") if x >= width || y >= height
+    @bits[y * width + x]
+  end
+
+  def [](xs : Range(Int32, Int32) | Range(Int32, Nil), y : Int32) : BitArray
+    start, count = resolve_to_start_and_count(xs, width)
+    BitArray.new(count) { |x| self[x + start, y] }
+  end
+
+  def [](x : Int32, ys : Range(Int32, Int32) | Range(Int32, Nil)) : BitArray
+    start, count = resolve_to_start_and_count(ys, height)
+    BitArray.new(count) { |y| self[x, y + start] }
+  end
+
+  def [](xs : Range(Int32, Int32) | Range(Int32, Nil), ys : Range(Int32, Int32) | Range(Int32, Nil)) : Array(BitArray)
+    start_x, count_x = resolve_to_start_and_count(xs, width)
+    start_y, count_y = resolve_to_start_and_count(ys, height)
+    count_y.times.to_a.map do |y|
+      BitArray.new(count_x) do |x|
+        self[x + start_x, y + start_y]
+      end
+    end
+  end
+
+  def ==(other : Mask)
+    width == other.width &&
+      bits == other.bits
+  end
+
+  def set(x : Int32, y : Int32, value : Bool) : Bool
+    raise IndexError.new("Out of bounds: this mask is #{width}x#{height}, and (#{x},#{y}) is outside of that") if x >= width || y >= height
+    @bits[y * width + x] = value
+  end
+
+  def []=(x : Int32, y : Int32, value : Bool) : Bool
+    self.set(x, y, value)
+  end
+
+  def []=(xs : Range(Int32, Int32) | Range(Int32, Nil), y : Int32, value : Bool) : Bool
+    raise IndexError.new("Out of bounds: #{y} is beyond the bounds of this mask's height of #{height}") if y >= height
+    start_x, count_x = resolve_to_start_and_count(xs, width)
+    @bits.fill(value, y * width + start_x, count_x)
+    value
+  end
+
+  def []=(x : Int32, ys : Range(Int32, Int32) | Range(Int32, Nil), value : Bool) : Bool
+    raise IndexError.new("Out of bounds: #{x} is beyond the bounds of this mask's width of #{width}") if x >= width
+    start_y, count_y = resolve_to_start_and_count(ys, height)
+    count_y.times.to_a.each do |y|
+      set(x, y + start_y, value)
+    end
+    value
+  end
+
+  def []=(xs : Range(Int32, Int32) | Range(Int32, Nil), ys : Range(Int32, Int32) | Range(Int32, Nil), value : Bool) : Bool
+    # IMPL: check ranges
+    start_x, count_x = resolve_to_start_and_count(xs, width)
+    start_y, count_y = resolve_to_start_and_count(ys, height)
+    count_y.times.to_a.each do |y|
+      @bits.fill(value, (y + start_y) * width + start_x, count_x)
+    end
+    value
+  end
+
+  def to_gray
+    GrayscaleImage.new(bits.map { |b| b ? 255u8 : 0u8 }, width, height)
```

Comment:
Method to convert a mask to a `GrayscaleImage` where true bits are white and false bits are black. Useful for debugging what a mask is generating: `mask.to_gray.to_ppm(file)` (PS: I'm really missing a `save(filename)` or similar method. Thoughts also welcome on what is the preferred interface for something like that).